client/my-sites/controller.js

```diff
@@ -334,22 +338,22 @@ module.exports = {
 		} else {
 			// if sites has fresh data and siteId is invalid
 			// redirect to allSitesPath
-			if ( ! isRequestingSites( getState() ) ) {
+			if ( ! isRequestingSites( state ) ) {
 				return page.redirect( allSitesPath );
 			}
 
 			let waitingNotice;
 			let freshSiteId;
 			const selectOnSitesChange = () => {
 				// if sites have loaded, but siteId is invalid, redirect to allSitesPath
-				freshSiteId = getSiteId( getState(), siteFragment );
+				freshSiteId = getSiteId( state, siteFragment );
```

Comment:
The `selectOnSitesChange` callback is executed at a different time than the enclosing `siteSelection`. You need to call `getState()` again here to get the fresh state (it has changed in the meantime) instead of reusing the stale one.

I don't know why exactly this PR had to be reverted, but this stands out as an obvious bug that can trigger an infinite loop.

Reply:
Good 👀 @jsnajdr !

Reply:
Nice find, thanks!